File: src/cadforge/core/events.py
"""Event model for async agent execution.

Structured events flow from the agent through an asyncio.Queue
to the REPL renderer, decoupling generation from display.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from enum import Enum
from typing import Any


class EventType(Enum):
    """Types of events emitted during agent execution."""

    TEXT_DELTA = "text_delta"
    TEXT_DONE = "text_done"
    TOOL_USE_START = "tool_use_start"
    TOOL_RESULT = "tool_result"
    STATUS = "status"
    ERROR = "error"
    COMPLETION = "completion"
    CANCELLED = "cancelled"


@dataclass
class AgentEvent:
    """A single event emitted by the agent during execution."""

    type: EventType
    data: dict[str, Any] = field(default_factory=dict)

    @staticmethod
    def text_delta(text: str) -> AgentEvent:
        """A chunk of streamed text."""
        return AgentEvent(type=EventType.TEXT_DELTA, data={"text": text})

    @staticmethod
    def text_done(text: str = "") -> AgentEvent:
        """End of a text content block."""
        return AgentEvent(type=EventType.TEXT_DONE, data={"text": text})

    @staticmethod
    def tool_use_start(name: str, tool_id: str) -> AgentEvent:
        """A tool_use content block has started."""
        return AgentEvent(
            type=EventType.TOOL_USE_START,
            data={"name": name, "id": tool_id},
        )

    @staticmethod
    def tool_result(name: str, tool_id: str, result: dict[str, Any]) -> AgentEvent:
        """A tool has finished executing."""
        return AgentEvent(
            type=EventType.TOOL_RESULT,
            data={"name": name, "id": tool_id, "result": result},
        )

    @staticmethod
    def status(message: str) -> AgentEvent:
        """A status update (e.g. "Thinking...", "Executing Bash...")."""
        return AgentEvent(type=EventType.STATUS, data={"message": message})

    @staticmethod
    def completion(text: str, usage: dict[str, int] | None = None) -> AgentEvent:
        """Agent has finished — final text and usage stats."""
        return AgentEvent(
            type=EventType.COMPLETION,
            data={"text": text, "usage": usage or {}},
        )

    @staticmethod
    def cancelled() -> AgentEvent:
        """Agent execution was cancelled by the user."""
        return AgentEvent(type=EventType.CANCELLED)

    @staticmethod
    def error(message: str) -> AgentEvent:
        """An error occurred during agent execution."""
        return AgentEvent(type=EventType.ERROR, data={"message": message})

File: packages/shared/src/events.ts
/**
 * Agent event types — shared between Node CLI and Python backend.
 */

export enum EventType {
  /** Agent is processing / thinking */
  STATUS = 'status',
  /** Streaming text chunk from LLM */
  TEXT_DELTA = 'text_delta',
  /** Complete text block (non-streaming) */
  TEXT = 'text',
  /** Tool use started */
  TOOL_USE_START = 'tool_use_start',
  /** Tool use completed with result */
  TOOL_RESULT = 'tool_result',
  /** Agent finished processing the message */
  COMPLETION = 'completion',
  /** Error occurred */
  ERROR = 'error',
  /** Permission request for a tool */
  PERMISSION_REQUEST = 'permission_request',
  /** Permission response from user */
  PERMISSION_RESPONSE = 'permission_response',
  /** Mode change (agent, plan, ask) */
  MODE_CHANGE = 'mode_change',
  /** Pipeline step (designer/coder/renderer/judge) */
  PIPELINE_STEP = 'pipeline_step',
  /** Pipeline round counter */
  PIPELINE_ROUND = 'pipeline_round',
  /** Pipeline rendered image */
  PIPELINE_IMAGE = 'pipeline_image',
  /** Pipeline judge verdict */
  PIPELINE_VERDICT = 'pipeline_verdict',
  /** Design status updated */
  DESIGN_UPDATED = 'design_updated',
  /** Design iteration saved */
  ITERATION_SAVED = 'iteration_saved',
  /** Design learnings indexed into vault */
  LEARNINGS_INDEXED = 'learnings_indexed',

  // ── Competitive pipeline events ──

  /** Competitive pipeline overall status change */
  COMPETITIVE_STATUS = 'competitive_status',
  /** Competitive round counter */
  COMPETITIVE_ROUND = 'competitive_round',
  /** Supervisor parsed specification */
  COMPETITIVE_SUPERVISOR = 'competitive_supervisor',
  /** Individual proposal status/progress */
  COMPETITIVE_PROPOSAL = 'competitive_proposal',
  /** Debate/critique exchange */
  COMPETITIVE_DEBATE = 'competitive_debate',
  /** Sandbox evaluation result */
  COMPETITIVE_SANDBOX = 'competitive_sandbox',
  /** Fidelity score result */
  COMPETITIVE_FIDELITY = 'competitive_fidelity',
  /** Merger/selector decision */
  COMPETITIVE_MERGER = 'competitive_merger',
  /** Learning extraction complete */
  COMPETITIVE_LEARNING = 'competitive_learning',
  /** Human approval requested — pipeline paused */
  COMPETITIVE_APPROVAL_REQUESTED = 'competitive_approval_requested',
  /** Human approval response received */
  COMPETITIVE_APPROVAL_RESPONSE = 'competitive_approval_response',
}

export interface AgentEvent {
  type: EventType;
  data: Record<string, unknown>;
  timestamp?: string;
}

export function makeEvent(type: EventType, data: Record<string, unknown>): AgentEvent {
  return { type, data, timestamp: new Date().toISOString() };
}
